src/hotspot/share/prims/jvmtiTagMapTable.cpp

```diff
@@ -37,190 +37,103 @@
-  entry->set_tag(tag);
-  return entry;
+oop JvmtiTagMapEntry::object() const {
+  assert(_obj == nullptr, "Must have a handle and not object");
+  return _wh.resolve();
 }
 
-void JvmtiTagMapTable::free_entry(JvmtiTagMapEntry* entry) {
-  entry->literal().release(JvmtiExport::weak_tag_storage()); // release to OopStorage
-  BasicHashtable<mtServiceability>::free_entry(entry);
+oop JvmtiTagMapEntry::object_no_keepalive() const {
+  assert(_obj == nullptr, "Must have a handle and not object");
+  return _wh.peek();
 }
 
-unsigned int JvmtiTagMapTable::compute_hash(oop obj) {
-  assert(obj != NULL, "obj is null");
-  return obj->identity_hash();
+JvmtiTagMapTable::JvmtiTagMapTable()
+  :_rrht_table(Constants::_table_size){
 }
 
-JvmtiTagMapEntry* JvmtiTagMapTable::find(int index, unsigned int hash, oop obj) {
-  assert(obj != NULL, "Cannot search for a NULL object");
-
-  for (JvmtiTagMapEntry* p = bucket(index); p != NULL; p = p->next()) {
-    if (p->hash() == hash) {
-
-      // Peek the object to check if it is the right target.
-      oop target = p->object_no_keepalive();
-
-      // The obj is in the table as a target already
-      if (target == obj) {
-        ResourceMark rm;
-        log_trace(jvmti, table)("JvmtiTagMap entry found for %s index %d",
-                                obj->print_value_string(), index);
-        // The object() accessor makes sure the target object is kept alive before
-        // leaking out.
-        (void)p->object();
-        return p;
-      }
+void JvmtiTagMapTable::clear() {
+  struct RemoveAll{
+    bool do_entry(JvmtiTagMapEntry   & entry, jlong const &  tag)
+    {
+      return true;
     }
-  }
-  return NULL;
-}
+  }RemoveAll;
+  _rrht_table.unlink(&RemoveAll);
 
-JvmtiTagMapEntry* JvmtiTagMapTable::find(oop obj) {
-  if (obj->fast_no_hash_check()) {
-    // Objects in the table all have a hashcode.
-    return NULL;
-  }
-  unsigned int hash = compute_hash(obj);
-  int index = hash_to_index(hash);
-  return find(index, hash, obj);
+  assert(_rrht_table.number_of_entries() == 0, "should have removed all entries");
 }
 
-JvmtiTagMapEntry* JvmtiTagMapTable::add(oop obj, jlong tag) {
-  unsigned int hash = compute_hash(obj);
-  int index = hash_to_index(hash);
-  // One was added while acquiring the lock
-  assert(find(index, hash, obj) == NULL, "shouldn't already be present");
-
-  // obj was read with AS_NO_KEEPALIVE, or equivalent.
-  // The object needs to be kept alive when it is published.
-  Universe::heap()->keep_alive(obj);
-
-  WeakHandle w(JvmtiExport::weak_tag_storage(), obj);
-  JvmtiTagMapEntry* p = new_entry(hash, w, tag);
-  Hashtable<WeakHandle, mtServiceability>::add_entry(index, p);
-  ResourceMark rm;
-  log_trace(jvmti, table)("JvmtiTagMap entry added for %s index %d",
-                          obj->print_value_string(), index);
-
-  // Resize if the table is getting too big.
-  resize_if_needed();
+JvmtiTagMapTable::~JvmtiTagMapTable() {
+  clear();
```

Comment:
Is this the only use of `clear`? If so we can just inline its code here and remove it from the API.

Reply:
This issue is not resolved.

Reply:
For clarity `clear()` is also called from `JvmtiTagMap::clear()`